src/H5Tpkg.h

```diff
@@ -479,357 +487,472 @@ H5_DLL herr_t H5T__commit_named(const H5G_loc_t *loc, const char *name, H5T_t *d
 H5_DLL H5T_t *H5T__open_name(const H5G_loc_t *loc, const char *name);
 H5_DLL hid_t  H5T__get_create_plist(const H5T_t *type);
 
+/* Helper function for H5T_convert that accepts a pointer to a H5T_conv_ctx_t structure */
+H5_DLL herr_t H5T_convert_with_ctx(H5T_path_t *tpath, H5T_t *src_type, H5T_t *dst_type,
```

Comment:
This is a new helper function for `H5T_convert` that helps avoid a severe amount of overhead during compound and other container datatype conversions where `H5T_convert` is called for every single member of every element. With the changes in this PR, the `H5T_conv_ctx_t` structure gets initialized in `H5T_convert`, meaning that any conversion function calling `H5T_convert` in a loop would initialize a new structure for every member of every element. In the case where an application-registered conversion exception function is being used, this would cause the conversion function to register and then tear down an ID for the source and destination datatypes each time, which was very expensive. Container type conversion functions can instead copy the current `H5T_conv_ctx_t` structure into a temporary one, update its `src_type_id` and `dst_type_id` fields with the datatype IDs for the current member being converted, and then call `H5T_convert_with_ctx` to avoid all the overhead.